migrations/src/main/scala/za/co/absa/enceladus/migrations/continuous/migrate01/EntityRepository.scala

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2018-2019 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.migrations.continuous.migrate01
+
+import org.mongodb.scala.MongoDatabase
+import org.mongodb.scala.bson.{BsonDocument, ObjectId}
+import org.mongodb.scala.model.Filters
+import org.mongodb.scala.model.Filters.{and, equal, or}
+import org.mongodb.scala.model.Sorts._
+import za.co.absa.atum.model.RunState
+import za.co.absa.enceladus.migrations.framework.ObjectIdTools
+
+/**
+  * The class contains a set of MongoDB versioned collection manipulation routines needed for continuous migration.
+  */
+final class EntityRepository(db: MongoDatabase, collectionName: String) {
+
+  import za.co.absa.enceladus.migrations.framework.dao.ScalaMongoImplicits._
+
+  /**
+    * Returns iterator to a sorted collection of documents. Documents are sorted by name and version.
+    *
+    * @return An iterator to a JSON representation of the documents.
+    */
+  def getSortedDocuments: Iterator[String] = {
+    db.getCollection(collectionName)
+      .find()
```

Comment:
Here we sort and iterate an entire collection on each call of continuous migration. Is it possible to iterate over the new documents which were added since the last run? As far as I understand, by default, [ObjectId](https://docs.mongodb.com/manual/reference/method/ObjectId/) in mongo includes a timestamp of when the document was created. Can we use it to filter the documents (ObjectId.getTimestamp() >= timestampOfTheLastMigratedDocumentInThePreviousRun) in the old instance? 

This could require special treatment in case of updatable objects, like runs. Possibly, we can have a moving timeHorizon: all the documents created before timeHorizon are safe to consider migrated and unchanged. The timeHorizon increases as we migrate documents, until we encounter an unfinished run (runs recognized to be in a Zombie state can be ignored).

What do you think?

Reply:
This is a good idea. Just wondering if it worth the effort. We can experiment how much time a continuous migration cycle takes in UAT on prod-like volume of data and decide if we need to optimize.

The last immutable state timestamp can be derived for sure. And we can store it in a nearby collection. But can we query MongoDb for the records arrived since the last timestamp?

Reply:
> Just wondering if it worth the effort. We can experiment how much time a continuous migration cycle takes in UAT on prod-like volume of data and decide if we need to optimize.

Right, it is a good idea to check on prod db dump. If it is not an issue right now, we can log it as a separate feature for future, when the database grows larger.  

>But can we query MongoDb for the records arrived since the last timestamp?

Yes. For example in mongo shell you can find (and count) all the objects created after June 1st 2019:

```
db.getCollection('dataset_v1').find(
  {_id:
    {$gt: ObjectId("5cf1a3e00000000000000000")}
  }).count()
```
Here I used this [calculator](https://steveridout.github.io/mongo-object-time/) to encode the timestamp (June 1st 2019) into ObjectId string.

Reply:
Thanks! Good to know. Going to check how much time it takes to run a continuous migration cycle, and then decide if we should implement the timestamp-based optimization.

Reply:
@yruslan Any update on this?

Reply:
Haven't checked it on a real database yet.

Reply:
Just ran on my local copy of a database. The migration runs very quickly (a couple of seconds).
The only thing to keep in mind is that indexes must exist for the old database since `sort` relies on it. If `sort` requires more than 32 MB the process will fail.

Reply:
Great to know! Thanks!